src/services/item/plugins/html/service.ts

```diff
@@ -0,0 +1,196 @@
+    uploadPath: string,
+    log?: FastifyBaseLogger,
+  ): Promise<Array<string>> {
+    if (!member) {
+      throw new UnauthorizedMember(member);
+    }
+
+    const children = await readdir(folder);
+
+    // we will flatMap with promises: first map
+    const uploads = children.map(async (child) => {
+      const childPath = path.join(folder, child);
+      const childUploadPath = path.join(uploadPath, child);
+      const stats = await lstat(childPath);
+
+      if (stats.isDirectory()) {
+        // recursively upload child folder
+        return await this.upload(member, childPath, childUploadPath, log);
+      } else {
+        // ignore this file if extension is not allowed
+        const ext = path.extname(childPath);
+
+        if (!this.validator.isExtensionAllowed(ext)) {
+          log?.info('HTML import: illegal file extension detected, skipping file: ', ext);
+          // we're using flatMap, represent none value with empty array
+          return [];
+        }
+
+        const mimetype = mime.getType(ext) ?? DEFAULT_MIME_TYPE;
+        await this.fileService.upload(member, {
+          file: fs.createReadStream(childPath),
+          filepath: childUploadPath,
+          mimetype,
+        });
+
+        // we're using flatMap, wrap result in array
+        return [childUploadPath];
+      }
+    });
+    // then resolve promises array and flatten
+    return (await Promise.all(uploads)).flat();
+  }
+
+  async createItem(
+    actor: Member,
+    repositories: Repositories,
+    htmlFile: MultipartFile,
+    onComplete: (
+      actor: Member,
+      baseName: string,
+      contentId: string,
+      parentId?: Item['id'],
+    ) => Promise<Item>,
+    parentId?: Item['id'],
+    log?: FastifyBaseLogger,
+  ): Promise<Item> {
+    /*
+      // uppy tries to guess the mime type but fails and falls back to application/octet-stream, so we disable this check for now
+      if (htmlFile.mimetype !== H5P_FILE_MIME_TYPE) {
+        throw new InvalidH5PFileError(htmlFile.mimetype);
+      }
+      */
+
+    const contentId = v4();
+    const tmpDir = await tmp.dir({ tmpdir: this.tempDir, unsafeCleanup: true });
+    const targetFolder = path.join(tmpDir.path, contentId);
+    const remoteRootPath = this.buildRootPath(this.pathPrefix, contentId);
+
+    await mkdir(targetFolder, { recursive: true });
+    const baseName = path.basename(htmlFile.filename, `.${this.extension}`);
+
+    // try-catch block for local storage cleanup
+    try {
+      const savePath = this.buildPackagePath(targetFolder, baseName);
+      const contentFolder = this.buildContentPath(targetFolder);
+
+      // save html file
+      await pipeline(htmlFile.file, fs.createWriteStream(savePath));
+      await extract(savePath, { dir: contentFolder });
+
+      const result = await this.validator.validatePackage(contentFolder);
+      if (!result.isValid) {
+        throw new HtmlInvalidFileError((result as { error: string }).error);
+      }
```

Comment:
Not sure why the cast is required. You could simplify by not returning the errors but throwing them from inside `validatePackage` directly, I think my original API is needlessly complex.